docs/sources/developers/plugins/error-handling.md

```diff
@@ -0,0 +1,59 @@
++++
+title = "Error handling"
+type = "docs"
++++
+
+# Error handling
+
+This guide explains how to handle errors in plugins.
+
+## Provide usable defaults
+
+Allow the user to learn your plugin in small steps. By providing a usable default configuration, the user can get started right away, and you can avoid unnecessary error messages early on.
+
+For example, by selecting the first field of a expected type, the panel can display a visualization without any user configuration. If a user explicitly selects a field we use that one, otherwise we default to the first field of type `string` that we find:
+
+```ts
+const numberField = frame.fields.find(field =>
+  options.numberFieldName ? field.name === options.numberFieldName : field.type === FieldType.number
+);
+```
+
+## Display errors messages
+
+To display an error message to the user, `throw` an `Error` with the message you want to display:
+
+```ts
+throw new Error('An error occurred.');
```

Comment:
Should we really end exception messages with a period? It's normally not good style.

Reply:
This really needs to be addressed in the UX style guide so that we are consistent. Hey @sarlinska, do we have a rule for this?
If not, this thread has good discussion: https://ux.stackexchange.com/questions/18671/periods-at-the-end-of-a-sentence-in-alert-message/18673

MSG says to leave periods off of UI text, which this might count as: https://docs.microsoft.com/en-us/style-guide/punctuation/periods

Reply:
@oddlittlebird It's not a question of UX, but of programming conventions, so Microsoft's style guide and the linked StackExchange question are not relevant for the topic at hand. It is also not relevant for the UX style guide. Convention dictates that programming exceptions/errors should not be punctuated.  I couldn't find anything stating this for TypeScript, but [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) it's stated for Go for example. For as long as I've been programming (close to 24 years), I've never seen punctuated exceptions/errors, so this is a first for me.

When it comes to UX, I think instead (as I mention in another comment) that the front-end code should be responsible for displaying a punctuated error message to the user, but that's independent of the internal exception error strings.

Reply:
See my comment [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings). If the UI doesn't punctuate the error messages, it should be done there.

Reply:
Again, I agree completely with you, @aknuds1. However, this how error handling works at the moment, and the guide should reflect that. If, and when we decide to improve error handling for plugin, we'll update this guide.

Reply:
I still think we shouldn't teach plugin authors an anti-pattern and it should be fixed in Grafana itself.